crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -63,6 +67,238 @@ static const uint8_t kAESCBCCiphertext[64] = {
+    { sha512, kPlaintext, kOutput_sha512, AWSLC_APPROVED },
+    { sha512_256, kPlaintext, kOutput_sha512_256, AWSLC_APPROVED }
+};
+
+class EVP_MD_ServiceIndicatorTest : public testing::TestWithParam<DigestTestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, EVP_MD_ServiceIndicatorTest, testing::ValuesIn(kDigestTestVectors));
+
+TEST_P(EVP_MD_ServiceIndicatorTest, EVP_Ciphers) {
+  const DigestTestVector &digestTestVector = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+  bssl::ScopedEVP_MD_CTX ctx;
+  std::vector<uint8_t> digest(digestTestVector.md.length);
+  unsigned digest_len;
+
+  // Test running the EVP_Digest interfaces one by one directly, and check
+  // |EVP_DigestFinal_ex| for approval at the end.
+  ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), digestTestVector.md.func(), nullptr));
+  ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), digestTestVector.input, sizeof(digestTestVector.input)));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_DigestFinal_ex(ctx.get(), digest.data(), &digest_len)));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.data(), digest_len, digestTestVector.md.name));
+  }
+
+  // Test using the one-shot |EVP_Digest| function for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, EVP_Digest(digestTestVector.input, sizeof(digestTestVector.input),
+                                               digest.data(), &digest_len, digestTestVector.md.func(), nullptr));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.data(), digest_len, digestTestVector.md.name));
+  }
+
+  // Test using the one-shot API for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, digestTestVector.md.one_shot_func(digestTestVector.input, sizeof(digestTestVector.input), digest.data()));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.data(), digestTestVector.md.length, digestTestVector.md.name));
+  }
+}
+
+struct HMACTestVector {
+  // func is the hash function for HMAC to test.
+  const EVP_MD *(*func)(void);
+  // input is a NUL-terminated string to hash.
+  const uint8_t *input;
+  // expected_digest is the expected digest.
+  const uint8_t *expected_digest;
+  // expected to be approved or not.
+  const int expect_approved;
+} kHMACTestVectors[] = {
+    { EVP_sha1, kPlaintext, kHMACOutput_sha1, AWSLC_APPROVED },
+    { EVP_sha224, kPlaintext, kHMACOutput_sha224, AWSLC_APPROVED },
+    { EVP_sha256, kPlaintext, kHMACOutput_sha256, AWSLC_APPROVED },
+    { EVP_sha384, kPlaintext, kHMACOutput_sha384, AWSLC_APPROVED },
+    { EVP_sha512, kPlaintext, kHMACOutput_sha512, AWSLC_APPROVED },
+    { EVP_sha512_256, kPlaintext, nullptr, AWSLC_NOT_APPROVED }
```

Comment:
yes, I think we should still check the expected output even though it is not approved. It also avoids all the if statements.